Building the MPI ABI using clang fails due to undefined types

Attemmpting to buld: MPICH v4.3.0  with `--enable-mpi-abi`
Using: clang `v20.1.5` 


While building `c_binding_abi.c`  in particular blocks like

```c
...
#if defined(HAVE_PRAGMA_WEAK)
#pragma weak MPI_Comm_create_keyval = PMPI_Comm_create_keyval
#elif defined(HAVE_PRAGMA_HP_SEC_DEF)
#pragma _HP_SECONDARY_DEF PMPI_Comm_create_keyval  MPI_Comm_create_keyval
#elif defined(HAVE_PRAGMA_CRI_DUP)
#pragma _CRI duplicate MPI_Comm_create_keyval as PMPI_Comm_create_keyval
#elif defined(HAVE_WEAK_ATTRIBUTE)
int MPI_Comm_create_keyval(ABI_Comm_copy_attr_function *comm_copy_attr_fn,
                           ABI_Comm_delete_attr_function *comm_delete_attr_fn, int *comm_keyval,
                           void *extra_state)
                            __attribute__ ((weak, alias("PMPI_Comm_create_keyval")));
#endif
...
```

I hit many errors such as

```
End of search list.
src/binding/abi/c_binding_abi.c:19:28: error: unknown type name 'ABI_Comm_copy_attr_function'; did you mean 'MPI_Comm_copy_attr_function'?
   19 | int MPI_Comm_create_keyval(ABI_Comm_copy_attr_function *comm_copy_attr_fn,
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                            MPI_Comm_copy_attr_function
```

With GCC it seems to work as the config check for `HAVE_PRAGMA_WEAK` passes, while it fails for `clang` due to https://github.com/llvm/llvm-project/issues/56760.

Was wondering if anyone else has tried it and if I am missing something obvious, or where those `ABI_xxx` types are supposedely defined?

--- a/maint/local_python/binding_c.py
+++ b/maint/local_python/binding_c.py
@@ -1148,7 +1148,7 @@ def dump_abi_wrappers(func, is_large):
                 return True
         return False
     # ----
-    re_Handle = r'MPI_(Comm|Datatype|Errhandler|Group|Info|Message|Op|Request|Session|Win|File)\b'
+    re_Handle = get_abi_re_Handle()
     for p in func['c_parameters']:
         skip_abi_swap = False
         param_type = mapping[p['kind']]
@@ -2959,6 +2959,10 @@ def dump_validate_get_topo_size(func):
 
 # ---- supporting routines (reusable) ----
 
+def get_abi_re_Handle():
+    # the following handle types need perform ABI-swap for abi bindings
+    return r'MPI_(Comm|Datatype|Errhandler|Group|Info|Message|Op|Request|Session|Win|File)\b'
+
 def is_type_create(func):
     last_p = func['c_parameters'][-1]
     return (last_p['kind'] == 'DATATYPE' and last_p['param_direction'] == 'out')
@@ -2974,16 +2978,10 @@ def get_declare_function(func, is_large, kind=""):
     name = get_function_name(func, is_large)
     mapping = get_kind_map('C', is_large)
 
-    ret = "int"
-    if 'return' in func:
-        ret = mapping[func['return']]
-        if func['return'] == 'EXTRA_STATE':
-            ret = 'void *'
+    ret = get_C_return(func, mapping, kind == 'abi')
 
-    params = get_C_params(func, mapping)
+    params = get_C_params(func, mapping, kind == 'abi')
     s_param = ', '.join(params)
-    if kind == 'abi':
-        s_param = re.sub(r'\bMPI_', 'ABI_', s_param)
     s = "%s %s(%s)" % (ret, name, s_param)
 
     if kind == 'proto':
@@ -2993,10 +2991,25 @@ def get_declare_function(func, is_large, kind=""):
         s += " MPICH_API_PUBLIC"
     return s
 
-def get_C_params(func, mapping):
+def get_C_return(func, mapping, is_abi=False):
+    ret = "int"
+    if 'return' in func:
+        ret = mapping[func['return']]
+        if func['return'] == 'EXTRA_STATE':
+            ret = 'void *'
+        if is_abi:
+            re_Handle = get_abi_re_Handle()
+            ret = re.sub(re_Handle, r'ABI_\1', ret)
+    return ret
+
+def get_C_params(func, mapping, is_abi=False):
     param_list = []
+    re_Handle = get_abi_re_Handle()
     for p in func['c_parameters']:
-        param_list.append(get_C_param(p, func, mapping))
+        param = get_C_param(p, func, mapping)
+        if is_abi:
+            param = re.sub(re_Handle, r'ABI_\1', param)
+        param_list.append(param)
     if not len(param_list):
         return ["void"]
     else:

--- a/maint/local_python/binding_c.py
+++ b/maint/local_python/binding_c.py
@@ -2997,7 +2997,7 @@ def get_C_return(func, mapping, is_abi=False):
         ret = mapping[func['return']]
         if func['return'] == 'EXTRA_STATE':
             ret = 'void *'
-        if is_abi:
+        if is_abi and func['dir'] != 'io':
             re_Handle = get_abi_re_Handle()
             ret = re.sub(re_Handle, r'ABI_\1', ret)
     return ret
@@ -3007,7 +3007,7 @@ def get_C_params(func, mapping, is_abi=False):
     re_Handle = get_abi_re_Handle()
     for p in func['c_parameters']:
         param = get_C_param(p, func, mapping)
-        if is_abi:
+        if is_abi and func['dir'] != 'io':
             param = re.sub(re_Handle, r'ABI_\1', param)
         param_list.append(param)
     if not len(param_list):

